src/FSharpPlus/Extensions/Array.fs

```diff
@@ -101,3 +101,12 @@ module Array =
     /// <returns>Array with corresponding pairs of input arrays.</returns>
     let zipShortest (a1: array<'T1>) (a2: array<'T2>) =
         Array.init (min a1.Length a2.Length) (fun i -> a1.[i], a2.[i])
+
+    /// <summary>Applies given function to each value of the given Array.</summary>
+    /// <param name="f">The mapping function.</param>
+    /// <param name="x">The input array.</param>
+    ///
+    /// <returns>Array with values x for each Array value where the function returns Some(x).</returns>
+    let chooseIndex f l =
```

Comment:
F# Core Array.choose implementation is quite complicated, so I think it's a good idea to reuse it instead of rolling our own. I have come up with something like below. 
This implementation is simple and doesn't allocate 2 arrays, but it is tied to evaluation order (which may be ok?).
```fsharp
let funnyChoosei f a =
    let mutable i = ref -1 
    let fi x = 
        incr i
        f !i x
    Array.choose fi a
```
@pzelmanski , @gusty  
what do you think?

Reply:
This would work for list too

Reply:
I think it's great, but at the same time it's a bit dangerous if F# core changes internally their implementation, but to be fair this is not very likely to happen, we can add unit tests to guard about it.
Very smart !

Reply:
> it's a bit dangerous if F# core changes internally their implementation

Is it more dangerous than it was before?
What kind of tests would you like to have to be guarded for such cases?

Reply:
I've replaced implementation of both list and array choosei and added tests for both in extensions